tx_parser_test.go

```diff
@@ -0,0 +1,442 @@
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.Err, func(t *testing.T) {
+			_, err := ParseContractTransaction(tt.Payload())
+			if err == nil {
+				t.Fatal("expecting an error, got nil instead")
+			}
+			assert.Contains(t, err.Error(), fmt.Sprintf("contract: %s", tt.Err))
+		})
+	}
+}
+
+func TestParseBatchTransaction(t *testing.T) {
+	batch := validBatch(t)
+	payload := encodeBatch(batch)
+
+	batch2, err := ParseBatchTransaction(payload)
+	assert.NoError(t, err)
+	assert.Equal(t, batch, batch2)
+}
+
+func TestParseBatchTransaction_Errors(t *testing.T) {
+	tests := []struct {
+		Err     string
+		Payload func() []byte
+	}{
+		{
+			"failed to decode number of transactions in batch",
+			func() []byte {
+				return []byte{}
+			},
+		},
+		{
+			"size must be greater than zero",
+			func() []byte {
+				return encodeBatch(Batch{})
+			},
+		},
+		{
+			"could not read tag",
+			func() []byte {
+				payload := encodeBatch(validBatch(t))
+				return payload[:1]
+			},
+		},
+		{
+			"entries inside batch cannot be batch transactions themselves",
+			func() []byte {
+				batch := validBatch(t)
+				batch.Tags[0] = uint8(sys.TagBatch)
+				batch.Payloads[0] = encodeBatch(batch)
+
+				return encodeBatch(batch)
+			},
+		},
+		{
+			"could not read payload size",
+			func() []byte {
+				payload := encodeBatch(validBatch(t))
+				return payload[:1+1+3]
+			},
+		},
+		{
+			"payload size exceeds 2MB",
+			func() []byte {
+				batch := validBatch(t)
+				batch.Payloads[0] = make([]byte, 2*1024*1024+1)
+
+				return encodeBatch(batch)
+			},
+		},
+		{
+			"could not read payload",
+			func() []byte {
+				payload := encodeBatch(validBatch(t))
+				return payload[:1+1+4+1]
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.Err, func(t *testing.T) {
+			_, err := ParseBatchTransaction(tt.Payload())
+			if err == nil {
+				t.Fatal("expecting an error, got nil instead")
+			}
+			assert.Contains(t, err.Error(), fmt.Sprintf("batch: %s", tt.Err))
+		})
+	}
+}
+
+func validTransfer(t *testing.T) Transfer {
+	keys, err := skademlia.NewKeys(sys.SKademliaC1, sys.SKademliaC2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return Transfer{
+		Recipient:  keys.PublicKey(),
+		Amount:     1337,
+		GasLimit:   42,
+		FuncName:   []byte("helloworld"),
+		FuncParams: []byte("foobar"),
+	}
+}
+
+func validStake(t *testing.T, opcode byte) Stake {
+	return Stake{
+		Opcode: opcode,
+		Amount: uint64(1337),
+	}
+}
+
+func validContract(t *testing.T) Contract {
+	return Contract{
+		GasLimit: 42,
+		Params:   []byte("foobar"),
+		Code:     []byte("loremipsumdolorsitamet"),
+	}
+}
+
+func validBatch(t *testing.T) Batch {
+	return Batch{
+		Size: 3,
+		Tags: []uint8{
+			uint8(sys.TagTransfer), uint8(sys.TagStake), uint8(sys.TagContract),
+		},
+		Payloads: [][]byte{
+			encodeTransfer(validTransfer(t)),
+			encodeStake(validStake(t, sys.PlaceStake)),
+			encodeContract(validContract(t)),
+		},
+	}
+}
+
+func encodeTransfer(tx Transfer) []byte {
```

Comment:
I think these `encode*` functions can be converted into member function of the payload struct, e.g.:

```go
func (t Transfer) Marshal() []byte {
    // blah
}
```

This would allow `wavelet/cmd` to be simplified, e.g.:

```go

func (cli *CLI) pay(cmd []string) {
	if len(cmd) != 2 {
		fmt.Println("pay <recipient> <amount>")
		return
	}

	recipient, err := hex.DecodeString(cmd[0])
	if err != nil {
		cli.logger.Error().Err(err).Msg("The recipient you specified is invalid.")
		return
	}

	if len(recipient) != wavelet.SizeAccountID {
		cli.logger.Error().Int("length", len(recipient)).Msg("You have specified an invalid account ID to find.")
		return
	}

	amount, err := strconv.ParseUint(cmd[1], 10, 64)
	if err != nil {
		cli.logger.Error().Err(err).Msg("Failed to convert payment amount to a uint64.")
		return
	}

	snapshot := cli.ledger.Snapshot()

	var recipientID wavelet.AccountID
	copy(recipientID[:], recipient)

	balance, _ := wavelet.ReadAccountBalance(snapshot, cli.keys.PublicKey())
	_, codeAvailable := wavelet.ReadAccountContractCode(snapshot, recipientID)

	if balance < amount+sys.TransactionFeeAmount {
		cli.logger.Error().Uint64("your_balance", balance).Uint64("amount_to_send", amount).Msg("You do not have enough PERLs to send.")
		return
	}

        payload := wavelet.Transfer{
            Recipient: recipient,
            Amount: amount,
        }

	if codeAvailable {
                payload.GasLimit = balance-sys.TransactionFeeAmount
                payload.FuncName = []byte("on_money_received")
	}

	tx, err := cli.sendTransaction(wavelet.NewTransaction(cli.keys, sys.TagTransfer, payload.Marshal()))
	if err != nil {
		return
	}

	cli.logger.Info().Msgf("Success! Your payment transaction ID: %x", tx.ID)
}
```

Maybe we can also introduce helper methods for creating each transaction type, e.g.:

```go
// This will simply call wavelet.NewTransaction(keys, sys.TagTransfer, payload.Marshal())
tx, err := wavelet.NewTransferTransaction(keys, wavelet.Transfer{Recipient: recipient, Amount: 1234})
```

Reply:
Completely agree on that - that sounds great :smile:.

Reply:
It will also affect `ParseJSON` since it is also building the payload. I think it's better to do it in a separate PR since there will be a lot of changes.